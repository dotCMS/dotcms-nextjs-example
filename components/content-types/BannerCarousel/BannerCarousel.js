import React, { useState } from 'react';
import { Carousel, CarouselItem, CarouselControl } from 'reactstrap';

const BannerCarousel = (props) => {
    const [activeIndex, setActiveIndex] = useState(0);

    let animating = false;

    const onExiting = () => {
        animating = true;
    };

    const onExited = () => {
        animating = false;
    };

    const next = () => {
        if (animating) return;
        const nextIndex = activeIndex === props.items.length - 1 ? 0 : activeIndex + 1;
        setActiveIndex(nextIndex);
    };

    const previous = () => {
        if (animating) return;
        const nextIndex = activeIndex === 0 ? props.items.length - 1 : activeIndex - 1;
        setActiveIndex(nextIndex);
    };

    const slides = props.items
        ? props.items.map((item) => {
              return (
                  <CarouselItem onExiting={onExiting} onExited={onExited} key={item.identifier}>
                      <div
                          className="banner bg-image-full bg-overlay-30 context-dark"
                          style={{
                              backgroundImage: `url(${`https://starter.dotcms.com`}/${item.image}/40q)`
                          }}
                      >
                          <div className="container">
                              <div className="row justify-content-lg-center">
                                  <div className="col-lg-9 text-center">
                                      <p className="banner-title">{item.title}</p>
                                      <h2 className="text-decoration-lines-2">
                                          <span>
                                              {item.caption}
                                              <span className="text-decoration-line text-decoration-line-left"></span>
                                              <span className="text-decoration-line text-decoration-line-right"></span>
                                          </span>
                                      </h2>
                                      {item.buttonText && item.link ? (
                                          <div className="group-lg">
                                              <a
                                                  className="button button-primary button-leaf"
                                                  href={item.link}
                                              >
                                                  {item.buttonText}
                                              </a>
                                          </div>
                                      ) : (
                                          ''
                                      )}
                                  </div>
                              </div>
                          </div>
                      </div>
                  </CarouselItem>
              );
          })
        : null;

    return (
        <Carousel
            className="carousel-banner"
            activeIndex={activeIndex}
            next={next}
            previous={previous}
        >
            {slides}
            <CarouselControl direction="prev" directionText="Previous" onClickHandler={previous} />
            <CarouselControl direction="next" directionText="Next" onClickHandler={next} />
        </Carousel>
    );
};

export default BannerCarousel;
