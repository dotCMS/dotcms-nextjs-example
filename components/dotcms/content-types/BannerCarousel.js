import React from 'react';
import htmlParser from '../../../utilities/htmlParser';
import { processNodes } from 'react-html-parser';
import { Carousel } from 'react-responsive-carousel';
import 'react-responsive-carousel/lib/styles/carousel.min.css'; // requires a loader

function BannerCarousel({ rendered }) {
    const transform = (node) => {
        if (node.type === 'tag') {
            if (node.attribs.class === 'carousel-inner') {
                return <Carousel>{processNodes(node.children)}</Carousel>;
            }

            if (node.attribs.class === 'carousel-control-prev') {
                return null;
            }

            if(node.attribs.class === "carousel-control-next") {
              return null;
            }      
        }
    };

    return htmlParser({ content: rendered, transform });
}

export default BannerCarousel;
