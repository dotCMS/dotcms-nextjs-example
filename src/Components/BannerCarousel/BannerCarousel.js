import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Parser from 'html-react-parser';
import { Carousel, CarouselItem, CarouselControl } from 'reactstrap';

class BannerCarousel extends Component {
    constructor(props) {
        super(props);
        this.state = { activeIndex: 0 };
        this.next = this.next.bind(this);
        this.previous = this.previous.bind(this);
        this.goToIndex = this.goToIndex.bind(this);
        this.onExiting = this.onExiting.bind(this);
        this.onExited = this.onExited.bind(this);
    }

    onExiting() {
        this.animating = true;
    }

    onExited() {
        this.animating = false;
    }

    next() {
        if (this.animating) return;
        const nextIndex =
            this.state.activeIndex === this.props.items.length - 1
                ? 0
                : this.state.activeIndex + 1;
        this.setState({ activeIndex: nextIndex });
    }

    previous() {
        if (this.animating) return;
        const nextIndex =
            this.state.activeIndex === 0
                ? this.props.items.length - 1
                : this.state.activeIndex - 1;
        this.setState({ activeIndex: nextIndex });
    }

    goToIndex(newIndex) {
        if (this.animating) return;
        this.setState({ activeIndex: newIndex });
    }

    render() {
        const displayedFields = this.props.fieldsToDisplay.split(',');
        const { activeIndex } = this.state;
        const slides = this.props.items.map(item => {
            const showTitle = displayedFields.includes('title');
            const showCaption = displayedFields.includes('caption1');
            return (
                <CarouselItem
                    onExiting={this.onExiting}
                    onExited={this.onExited}
                    key={item.src}
                >
                    <img src={item.src} alt={item.altText} />
                    {showTitle || showCaption ? (
                        <div className="carousel-caption d-none d-md-block">
                            {showTitle ? <h1>{item.title}</h1> : ''}
                            {showCaption ? Parser(item.caption) : ''}
                        </div>
                    ) : (
                        ''
                    )}
                </CarouselItem>
            );
        });

        return (
            <Carousel
                activeIndex={activeIndex}
                next={this.next}
                previous={this.previous}
            >
                {slides}
                <CarouselControl
                    direction="prev"
                    directionText="Previous"
                    onClickHandler={this.previous}
                />
                <CarouselControl
                    direction="next"
                    directionText="Next"
                    onClickHandler={this.next}
                />
            </Carousel>
        );
    }
}

BannerCarousel.propTypes = {
    fieldsToDisplay: PropTypes.string.isRequired,
    items: PropTypes.array.isRequired
};

export default BannerCarousel;
