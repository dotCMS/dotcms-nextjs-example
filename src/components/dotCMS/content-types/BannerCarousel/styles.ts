// Dependencies
import styled from 'styled-components'

export const CarouselContainer = styled.div`
  .carousel-inner {
    position: relative;
    height: 300px;
    width: 100%;
  }

  .banner.bg-image-full {
    background-repeat: no-repeat;
    width: 100%;
    background-size: cover;
    height: 350px;
    display: flex;
    flex-direction: column;
    justify-content: center;
  }
  .row.justify-content-lg-center {
    display: flex;
    justify-content: center;
  }

  .banner-title {
    font-size: 2.25rem;
    line-height: 4rem;
    font-weight: bold;
    margin-bottom: 0;
  }

  .banner-title,
  .text-decoration-lines-2 {
    color: white;
  }

  .text-decoration-lines-2 {
    font-size: 1.2rem;
    line-height: 1.6rem;
  }
  .bg-overlay-30::before {
    z-index: 0;
    position: absolute;
    top: 0;
    left: 0;
    background: rgba(0, 0, 0, 0.3);
    width: 100%;
    height: 100%;
    content: '';
  }
`
