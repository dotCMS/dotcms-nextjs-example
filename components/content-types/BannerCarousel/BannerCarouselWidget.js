import React, { useState, useEffect } from 'react';
import BannerCarousel from './BannerCarousel';

const BannerCarouselWidget = props => {
    const [items, setItems] = useState([]);

    useEffect(() => {
        if (!items.length) {
            fetch(`${`https://starter.dotcms.com`}/api/content/id/${props.identifier}/depth/1`)
                .then(response => response.json())
                .then(({ contentlets }) => {
                    if (contentlets) {
                        setItems(contentlets[0].banners);
                        
                    }
                })
                .catch(error => console.error(error));
        }
    });

    return (
        items.length ? <BannerCarousel items={items} /> : ''
    );
};

export default BannerCarouselWidget;
