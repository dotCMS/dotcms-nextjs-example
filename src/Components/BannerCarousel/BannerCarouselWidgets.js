import React from 'react';
import DotCMSApi from '../../libs/dotcms.api';
import { Spinner } from 'reactstrap';
import NoResults from '../Shared/NoResults';
import BannerCarousel from './BannerCarousel';

const sortResultsByMap = {
    title: 'title_dotraw',
    modDate: 'modDate'
};

class BannerCarouselWidgets extends React.Component {
    constructor(props, context) {
        super(props, context);
        this.state = {
            loading: true,
            banners: []
        };
    }

    formatBannersData(banners) {
        return banners.map(banner => {
            return {
                src: banner.imageVersion,
                altText: banner.title,
                title: banner.title,
                caption: banner.caption1
            };
        });
    }

    getBannersItems() {
        const defaultEsParams = {
            languageId: this.props.data.languageId,
            numberOfResults: this.props.data.numberOfResults,
            sortOrder1: this.props.data.sortOrder1,
            sortResultsBy: sortResultsByMap[this.props.data.sortResultsBy]
        };

        DotCMSApi.esSearch('banner', defaultEsParams)
            .then(response => response.json())
            .then(bannerData => {
                this.setState(state => ({
                    ...state,
                    loading: false,
                    banners: this.formatBannersData(bannerData.contentlets)
                }));
            });
    }

    componentDidMount() {
        this.getBannersItems();
    }

    render() {
        const { loading, banners } = this.state;
        if (loading) {
            return <Spinner color="primary" />;
        } else {
            return banners && banners.length ? (
                <BannerCarousel
                    items={banners}
                    fieldsToDisplay={this.props.data.fieldsToDisplay}
                />
            ) : (
                <NoResults />
            );
        }
    }
}

export default BannerCarouselWidgets;
