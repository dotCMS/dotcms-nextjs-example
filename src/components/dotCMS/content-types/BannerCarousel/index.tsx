// Dependencies
import { domToReact } from 'html-react-parser'
import { nanoid } from 'nanoid'
import { Carousel } from 'react-responsive-carousel'
import 'react-responsive-carousel/lib/styles/carousel.min.css' // requires a loader
import type { HTMLReactParserOptions } from 'html-react-parser'
import type { Element } from 'html-react-parser'

// Internals
import { htmlParser } from '@/utils'
import { CarouselContainer } from './styles'

export type BannerCarouselProps = {
  rendered: string
}

export const BannerCarousel = ({ rendered }: BannerCarouselProps) => {
  const options: HTMLReactParserOptions = {
    replace: (domNode) => {
      const typedDomNode = domNode as Element
      const attribs = typedDomNode.attribs

      if (attribs?.className === 'carousel-inner') {
        return (
          <Carousel key={nanoid()} showThumbs={false}>
            {domToReact(typedDomNode.children) as JSX.Element[]}
          </Carousel>
        )
      }

      if (
        attribs?.className === 'carousel-control-prev' ||
        attribs?.className === 'carousel-control-next'
      ) {
        return null
      }
    },
  }

  return (
    <CarouselContainer>
      {htmlParser({
        content: rendered,
        options,
      })}
    </CarouselContainer>
  )
}

export default BannerCarousel
