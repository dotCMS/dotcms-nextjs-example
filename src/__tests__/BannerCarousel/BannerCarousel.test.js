import React from 'react';
import { shallow } from 'enzyme';
import BannerCarousel from '../../Components/BannerCarousel';
import toJSON from 'enzyme-to-json';

const bannersMock = [{
    src: 'url test',
    altText: 'alt text test',
    title: 'tile test',
    caption: 'caption test'
}];

describe('<BannerCarousel />', () => {
    it('renders banners carousel correctly', () => {
        const fieldsToDisplay = 'title,caption1';
        const wrapper = shallow(
            <BannerCarousel items={bannersMock} fieldsToDisplay={fieldsToDisplay} />
        );
        expect(toJSON(wrapper)).toMatchSnapshot();
    });

    it('renders banners, without "title"', () => {
        const fieldsToDisplay = 'caption1';
        const wrapper = shallow(
            <BannerCarousel items={bannersMock} fieldsToDisplay={fieldsToDisplay} />
        );
        expect(toJSON(wrapper)).toMatchSnapshot();
    });

    it('renders banners, without "caption"', () => {
        const fieldsToDisplay = 'title';
        const wrapper = shallow(
            <BannerCarousel items={bannersMock} fieldsToDisplay={fieldsToDisplay} />
        );
        expect(toJSON(wrapper)).toMatchSnapshot();
    });

    it('renders banners, without "title" & "caption"', () => {
        const fieldsToDisplay = '';
        const wrapper = shallow(
            <BannerCarousel items={bannersMock} fieldsToDisplay={fieldsToDisplay} />
        );
        expect(toJSON(wrapper)).toMatchSnapshot();
    });
});
