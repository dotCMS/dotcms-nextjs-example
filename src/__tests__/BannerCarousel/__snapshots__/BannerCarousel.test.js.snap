// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BannerCarousel /> renders banners carousel correctly 1`] = `
<Carousel
  activeIndex={0}
  interval={5000}
  keyboard={true}
  next={[Function]}
  pause="hover"
  previous={[Function]}
  slide={true}
>
  <CarouselItem
    appear={false}
    enter={true}
    exit={true}
    in={false}
    key="url test"
    mountOnEnter={false}
    onEnter={[Function]}
    onEntered={[Function]}
    onEntering={[Function]}
    onExit={[Function]}
    onExited={[Function]}
    onExiting={[Function]}
    slide={true}
    tag="div"
    timeout={600}
    unmountOnExit={false}
  >
    <img
      alt="alt text test"
      src="url test"
    />
    <div
      className="carousel-caption d-none d-md-block"
    >
      <h1>
        tile test
      </h1>
      caption test
    </div>
  </CarouselItem>
  <CarouselControl
    direction="prev"
    directionText="Previous"
    onClickHandler={[Function]}
  />
  <CarouselControl
    direction="next"
    directionText="Next"
    onClickHandler={[Function]}
  />
</Carousel>
`;

exports[`<BannerCarousel /> renders banners, without "caption" 1`] = `
<Carousel
  activeIndex={0}
  interval={5000}
  keyboard={true}
  next={[Function]}
  pause="hover"
  previous={[Function]}
  slide={true}
>
  <CarouselItem
    appear={false}
    enter={true}
    exit={true}
    in={false}
    key="url test"
    mountOnEnter={false}
    onEnter={[Function]}
    onEntered={[Function]}
    onEntering={[Function]}
    onExit={[Function]}
    onExited={[Function]}
    onExiting={[Function]}
    slide={true}
    tag="div"
    timeout={600}
    unmountOnExit={false}
  >
    <img
      alt="alt text test"
      src="url test"
    />
    <div
      className="carousel-caption d-none d-md-block"
    >
      <h1>
        tile test
      </h1>
    </div>
  </CarouselItem>
  <CarouselControl
    direction="prev"
    directionText="Previous"
    onClickHandler={[Function]}
  />
  <CarouselControl
    direction="next"
    directionText="Next"
    onClickHandler={[Function]}
  />
</Carousel>
`;

exports[`<BannerCarousel /> renders banners, without "title" & "caption" 1`] = `
<Carousel
  activeIndex={0}
  interval={5000}
  keyboard={true}
  next={[Function]}
  pause="hover"
  previous={[Function]}
  slide={true}
>
  <CarouselItem
    appear={false}
    enter={true}
    exit={true}
    in={false}
    key="url test"
    mountOnEnter={false}
    onEnter={[Function]}
    onEntered={[Function]}
    onEntering={[Function]}
    onExit={[Function]}
    onExited={[Function]}
    onExiting={[Function]}
    slide={true}
    tag="div"
    timeout={600}
    unmountOnExit={false}
  >
    <img
      alt="alt text test"
      src="url test"
    />
  </CarouselItem>
  <CarouselControl
    direction="prev"
    directionText="Previous"
    onClickHandler={[Function]}
  />
  <CarouselControl
    direction="next"
    directionText="Next"
    onClickHandler={[Function]}
  />
</Carousel>
`;

exports[`<BannerCarousel /> renders banners, without "title" 1`] = `
<Carousel
  activeIndex={0}
  interval={5000}
  keyboard={true}
  next={[Function]}
  pause="hover"
  previous={[Function]}
  slide={true}
>
  <CarouselItem
    appear={false}
    enter={true}
    exit={true}
    in={false}
    key="url test"
    mountOnEnter={false}
    onEnter={[Function]}
    onEntered={[Function]}
    onEntering={[Function]}
    onExit={[Function]}
    onExited={[Function]}
    onExiting={[Function]}
    slide={true}
    tag="div"
    timeout={600}
    unmountOnExit={false}
  >
    <img
      alt="alt text test"
      src="url test"
    />
    <div
      className="carousel-caption d-none d-md-block"
    >
      caption test
    </div>
  </CarouselItem>
  <CarouselControl
    direction="prev"
    directionText="Previous"
    onClickHandler={[Function]}
  />
  <CarouselControl
    direction="next"
    directionText="Next"
    onClickHandler={[Function]}
  />
</Carousel>
`;
